[wasm] Unmanaged structs are considered blittable if module has `DisableRuntimeMarshallingAttribute`

maraf opened the pull request:
Consider all types in pinvokes from assembly marked with `DisableRuntimeMarshallingAttribute` as blittable.
Based on API proposal in https://github.com/dotnet/runtime/issues/60639.

- PInvokeTableGenerator: Avoid crash when processing unmanaged callbacks with function pointers.

Fixes #61146.

maraf committed:
[wasm] Unmanaged structs are considered blittable if module has DisableRuntimeMarshallingAttribute.

maraf commented:
/azp run runtime-wasm

jkoritzinsky commented:
This is not entirely true. An unmanaged struct defined in an assembly with DisableRuntimeMarshalling isn't necessarily blittable if it is used in a P/Invoke in an assembly without DisableRuntimeMarshallingAttribute.

maraf committed:
Fix WBT.

maraf commented:
> This is not entirely true. An unmanaged struct defined in an assembly with DisableRuntimeMarshalling isn't necessarily blittable if it is used in a P/Invoke in an assembly without DisableRuntimeMarshallingAttribute.

Thanks!
Must both struct and p/invoke assemblies be marked with `DisableRuntimeMarshallingAttribute`?

jkoritzinsky commented:
Only the P/Invoke assembly must be marked with the attribute.

maraf reviewed (commented):
~~Look for the DisableRuntimeMarshallingAttribute on the assembly where pinvoke method is declared.~~
Mistake...

maraf committed:
Look for the DisableRuntimeMarshallingAttribute on the assembly where pinvoke method is declared.

maraf commented:
/azp run runtime-wasm

maraf committed:
Fix

maraf commented:
/azp run runtime-wasm

maraf commented:
/azp run runtime-wasm

maraf committed:
Merge remote-tracking branch 'upstream/main' into WasmPInvokeGeneratorDisableRuntimeMarshalling

maraf commented:
/azp run runtime-wasm

lambdageek commented on src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:
```diff
@@ -341,8 +341,12 @@ private static void EmitNativeToInterp(StreamWriter w, List<PInvokeCallback> cal
             MethodInfo method = cb.Method;
             bool isVoid = method.ReturnType.FullName == "System.Void";
 
+            if (method.DeclaringType != null && method.DeclaringType.Assembly.GetCustomAttributesData().Any(d => d.AttributeType.Name == "DisableRuntimeMarshallingAttribute"))
```
Probably doesn't matter too much, but maybe we can keep a cache for each assembly if it has the marshaling attribute?

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -115,6 +115,43 @@ public static int Main()
 
+        [Theory]
+        [BuildAndRun(host: RunHost.Chrome)]
+        public void UnmanagedStructsAreConsideredBlittable(BuildArgs buildArgs, RunHost host, string id)
```
And one variation of the test without the attribute.

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -115,6 +115,43 @@ public static int Main()
+                {
+                    var x = new S { Value = 5 };
+            
+                    Console.WriteLine("Main running " + x.Value);
```
This is using the value from the local struct. Can we call `M(..)`?

radical replied:
err.. get a callback?

maraf replied:
The error comes from the build. I'm using the struct it just don't get trimmed.

radical replied:
It would test whether we generated a good signature for it, right?

maraf replied:
Yes, when it succeeds it doesn't produce any warning.

maraf committed:
Feedback

- Cache DisableRuntimeMarshallingAttribute on assembly.
- Test without DisableRuntimeMarshallingAttribute.

maraf committed:
WasmBuildNative only without AOT.

maraf committed:
Merge remote-tracking branch 'upstream/main' into WasmPInvokeGeneratorDisableRuntimeMarshalling

maraf commented:
/azp run runtime-wasm

maraf commented:
Failures are not relevant.

radical commented on src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:
```diff
@@ -457,6 +461,16 @@ private static void EmitNativeToInterp(StreamWriter w, List<PInvokeCallback> cal
         w.WriteLine("};");
     }
 
+    private Dictionary<Assembly, bool> assemblyDisableRuntimeMarshallingAttributeCache = new Dictionary<Assembly, bool>();
```
```suggestion
    private Dictionary<Assembly, bool> _assemblyDisableRuntimeMarshallingAttributeCache = new ();
```

And move this to the top, with the other fields.

radical commented on src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:
```diff
@@ -457,6 +461,16 @@ private static void EmitNativeToInterp(StreamWriter w, List<PInvokeCallback> cal
+    private bool HasAssemblyDisableRuntimeMarshallingAttribute(Assembly assembly)
+    {
+        if (!assemblyDisableRuntimeMarshallingAttributeCache.TryGetValue(assembly, out var value))
+            assemblyDisableRuntimeMarshallingAttributeCache[assembly] = value = assembly.GetCustomAttributesData().Any(d => d.AttributeType.Name == "DisableRuntimeMarshallingAttribute");
```
```suggestion
            assemblyDisableRuntimeMarshallingAttributeCache[assembly]
            		= value = assembly.GetCustomAttributesData()
            								.Any(d => d.AttributeType.Name == nameof(DisableRuntimeMarshallingAttribute));
```

nit: Change the indentation here as you want, just preferably with shorter lines.

radical replied:
oops, I added tabs in that.

radical commented:
> Only the P/Invoke assembly must be marked with the attribute.

We need a test that covers this case also. The tests in the PR right now, have the struct, and the pinvoke are in the same assembly.

maraf committed:
Feedback.

maraf committed:
Merge remote-tracking branch 'upstream/main' into WasmPInvokeGeneratorDisableRuntimeMarshalling

maraf committed:
Unit test with struct in different assembly.

maraf commented:
/azp run runtime-wasm

radical commented on src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:
```diff
@@ -14,6 +14,7 @@
 internal sealed class PInvokeTableGenerator
 {
     private static readonly char[] s_charsToReplace = new[] { '.', '-', '+' };
+    private Dictionary<Assembly, bool> _assemblyDisableRuntimeMarshallingAttributeCache = new Dictionary<Assembly, bool>();
```
```suggestion
    private Dictionary<Assembly, bool> _assemblyDisableRuntimeMarshallingAttributeCache = new ();
```

maraf committed:
Update src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs

Co-authored-by: Ankit Jain <radical@gmail.com>

maraf commented:
/azp run runtime-wasm

maraf commented:
Failures doesn't seem relevant.

https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-73310-merge-6827ac8318034a2eb2/Workloads-.WasmTemplateTests/1/console.88ad2f22.log?helixlogtype=result

radical committed:
Merge remote-tracking branch 'origin/main' into WasmPInvokeGeneratorDisableRuntimeMarshalling

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -119,6 +119,198 @@ public static int Main()
+                    AssertAppBundle: false
+                )
+            );
+
```
```suggestion
```

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -119,6 +119,198 @@ public static int Main()
 
+        [Theory]
+        [BuildAndRun(host: RunHost.None)]
+        public void UnmanagedStructsAreNotConsideredBlittable(BuildArgs buildArgs, string id)
```
The 3 tests seem to have lot of the same code. That can be moved to a separate method.

radical replied:
Maybe `UnmanagedStructsWithoutDisableMarshallingAttributeAreNotConsideredBlittable`

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -119,6 +119,198 @@ public static int Main()
+
+            public struct S { public int Value; }
+
+            class Program
+            {
+                public static void Main()
+                {
+                    // Noop
+                }
+            }
```
This could be dropped, and built with `<OutputType>library</OutputType>`

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -119,6 +119,198 @@ public static int Main()
+
+        [Theory]
+        [BuildAndRun(host: RunHost.Chrome)]
+        public void UnmanagedStructsAreConsideredBlittable(BuildArgs buildArgs, RunHost host, string id)
```
The name should reflect what case this is, since there are multiple here.

The cases that should be covered IMHO, are:

1. Method, and struct in same assembly with no attribute
2. Separate assemblies - only Struct one has attribute
3. Separate assemblies - only method one has attribute
4. Separate assemblies - both have attribute
5. Separate assemblies - neither have the attribute

They are very similar, so having the test as a method which gets used by different cases, should make it easier to understand too.

radical replied:
I'll add them.

radical committed:
Merge remote-tracking branch 'origin/main' into WasmPInvokeGeneratorDisableRuntimeMarshalling

radical committed:
Update, and add some additional tests

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -119,6 +119,195 @@ public static int Main()
+                    new object[] { /*libraryHasAttribute*/ true, /*appHasAttribute*/ false, /*expectSuccess*/ false },
+                    new object[] { /*libraryHasAttribute*/ false, /*appHasAttribute*/ true, /*expectSuccess*/ true },
+                    new object[] { /*libraryHasAttribute*/ true, /*appHasAttribute*/ true, /*expectSuccess*/ true }
+                ).WithRunHosts(RunHost.V8).UnwrapItemsAsArrays();
```
```suggestion
                ).WithRunHosts(RunHost.Chrome).UnwrapItemsAsArrays();
```

radical commented on src/tests/BuildWasmApps/Wasm.Build.Tests/PInvokeTableGeneratorTests.cs:
```diff
@@ -119,6 +119,195 @@ public static int Main()
+        }
+
+        [Theory]
+        [BuildAndRun(host: RunHost.V8)]
```
```suggestion
        [BuildAndRun(host: RunHost.Chrome)]
```

radical committed:
Apply suggestions from code review

radical commented:
/azp run runtime-wasm

radical committed:
[wasm] PInvokeTableGenerator: Avoid crash when processing unmanaged

.. callbacks with function pointers.

radical committed:
Merge remote-tracking branch 'origin/main' into WasmPInvokeGeneratorDisableRuntimeMarshalling

radical commented:
/azp run runtime-wasm

radical commented:
The wasm EAT/AOT test failures are https://github.com/dotnet/runtime/issues/73932 .
The Perftracing build failure is https://github.com/dotnet/runtime/issues/73933 .

radical commented:
The Linux_musl x64 coreclr failure is https://github.com/dotnet/runtime/issues/73930 .

kasperk81 commented on src/tasks/WasmAppBuilder/PInvokeTableGenerator.cs:
```diff
@@ -460,6 +486,18 @@ private static void EmitNativeToInterp(StreamWriter w, List<PInvokeCallback> cal
+        {
+            _assemblyDisableRuntimeMarshallingAttributeCache[assembly] = value = assembly
+                .GetCustomAttributesData()
+                .Any(d => d.AttributeType.Name == "DisableRuntimeMarshallingAttribute");
```
can use `nameof(DisableRuntimeMarshallingAttribute)` with `using System.Runtime.CompilerServices;`

maraf replied:
The assembly is also compiled for net47 and the attribute is not available there.
